Add a timeout to fi_ud_pingpong test

jsquyres opened the pull request:
Since fi_ud_pingpong is a simple UD pingpong test, it has no retrans functionality.  Hence, if there's even a single drop, the test will hang (this causes Cisco's internal Jenkins' tests to arbitrarily "fail").

Instead, if any individual receive goes longer than [timeout] seconds, just quit the test and return the runfabtests.sh-understood 61 exit status to indicate that the test was skipped.

[ ***EDIT the original version of this PR returned value 77, not 61*** ]

@goodell Per our discussion yesterday, please review this.

goodell commented:
I know that `fi_ud_pingpong` isn't the best benchmark of our DGRAM performance, but won't this approach introduce noticeable overhead in the latency numbers?  What about using [`alarm(2)`](http://linux.die.net/man/2/alarm) instead?

Also, the usage of `clock_gettime` will probably cause problems on OS X, where I do not think it exists.

goodell commented on pingpong/ud_pingpong.c:
```diff
@@ -564,6 +577,10 @@ int main(int argc, char **argv)
+	// If the test timed out, return "61" so that runtests.sh will
+	// indicate that this test was (effectively) not run.
+	if (ret == -FI_ETIMEDOUT)
+		ret = -61;
```
FYI, `-61` is `-FI_ENODATA`

bturrubiates commented:
@goodell is correct about the use of `clock_gettime`, this test would not be able to be tested on OS X with the sockets provider. 

Using `alarm` sounds like a better solution.

shefty commented:
Btw, the runfabtests script will handle test timeout failures and abort tests.

shefty commented:
clock_gettime has been implemented for os-x.  It's already used by the complex fabtest app, which it does in order to timeout waiting for lost datagrams.  :)

bturrubiates commented:
Well yes `clock_gettime` itself has been implemented but on OS X does not support `CLOCK_REALTIME_COURSE`. That could be easily made `CLOCK_REALTIME` and then it wouldn't be a problem.

shefty commented:
Rather than aborting the test, why not just retry the transfer?

jsquyres commented:
We talked about adding retransmission to this test, but that seems to defeat the point of having a "simple" pingpong.  We have also talked about adding fi_ud_pingpong_with_retrans (or something like that) -- that may happen (but not part of this PR).

This PR is a "quick" way to get to "the test didn't fail, but it didn't pass, either."  More specifically, we want our Jenkins to *not* report a UD drop -- which currently causes this test to hang -- as a "fail."  We'd rather that this is reported as a "skip".

jsquyres commented:
Internally, @bturrubiates and @goodell suggested that I use `alarm(2)` instead of `clock_gettime(2)`.  I started looking into that, but don't have a conclusion here yet (I'm at the MPI Forum this week and am supposedly paying attention to the discussions here... :) ).

jsquyres commented:
FWIW, after doing some performance testing, using `clock_gettime(CLOCK_REALTIME_COARSE)` causes less of a performance disruption compared to `alarm(2)`.

Specifically, using `alarm(2)` at the beginning of each `recv_xfer()` cause a trap down to the OS, and results in noticeably lower bandwidth in the 64k and 1m tests.  Using `clock_gettime(C_RT_COARSE)` is purely a userspace operation in Linux, and results in exactly the same bandwidth in the 64k and 1m tests as having no timeout checking at all.

On OS X, it looks like we should not be using host_get_clock_service() (see http://stackoverflow.com/a/21352348/1120009) -- apparently, it's really slow and inaccurate.  Believe it or not, gettimeofday() is apparently highly optimized on OS X; we should probably just be using that...!

I'll submit an updated PR soon with these cleanups.

jsquyres commented:
I added some minor cleanups to the patches, filed a separate PR for the OS X `clock_gettime` emulation of `CLOCK_REALTIME_COARSE` (see #270), and also added these 2 minor fixes to runtests:

* avoid redirecting from a non-existent file
* also emit the number of notrun tests at the bottom

The runtests fixes came about organically while working on this stuff; I can separate them into a distinct PR if people want.

@goodell @bturrubiates @shefty Please review.

goodell commented:
> FWIW, after doing some performance testing, using clock_gettime(CLOCK_REALTIME_COARSE) causes less of a performance disruption compared to alarm(2).
>
> Specifically, using alarm(2) at the beginning of each recv_xfer() cause a trap down to the OS, and results in noticeably lower bandwidth in the 64k and 1m tests. Using clock_gettime(C_RT_COARSE) is purely a userspace operation in Linux, and results in exactly the same bandwidth in the 64k and 1m tests as having no timeout checking at all.

My thought had been that you would set the `alarm` at the beginning of either the whole test or at the beginning of each message size change, before the timer is started to compute performance.  That shouldn't result in any adverse performance impact, right?

bturrubiates commented:
@jsquyres I don't see the "2 minor fixes" that you added. Did you remove them from the PR?

jsquyres commented:
@bturrubiates See 1dc359b and 8c727bb (the last 2 commits on this PR).

@goodell I suppose.  Does it matter?  The short message latency doesn't appear to move with or without it.

goodell commented:
> Does it matter? The short message latency doesn't appear to move with or without it.

So long as you can't see a `usnic` provider small message latency difference with/without your patch, then I don't care.
